### PR TITLE
LP-2378 Backend implementation of user application cover letter page

### PR DESCRIPTION
**Pull request by annas-usman**
**urls.py** has been modified
**views.py** has been modified by adding a new class view

[Ticket LP-2378](https://philanthropyu.atlassian.net/browse/LP-2378)

[adg-edx-theme](https://github.com/OmnipreneurshipAcademy/adg-edx-theme/pull/76)

### CODE REVIEW

File: openedx/adg/lms/applications/views.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/views.py:231]: Do we need `get_or_create` here? Or will `get` suffice?
Also, we can access the ApplicationHub with related name on the user object as `user.application_hub`

**Review by MunamMubashir-arbisoft — CHANGES_REQUESTED** [review]

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [inline openedx/adg/lms/applications/views.py:208]: Do we need these checks when we are already doing front end validations, and making sure that business line is selected. A post request won't be triggered unless a business line is selected by the user. I think you wrote that logic in cover_letter.js `showpop()` function

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [inline openedx/adg/lms/applications/views.py:226]: Why are we calling `user_application.save()` multiple times? It is called in line 211 as well. According to the current implementation, a call to the db would be made twice, first to store the business line, and then to store the cover letter. Both the values should be first populated in memory and then a single call should be made to the db to store the updated object.

Comment (harris-irfan) [reply openedx/adg/lms/applications/views.py:226]: Moreover, do we really need to store user application data if the user click on the back button? According to the current implementation, even if the user click the 'Back' button on the confirmation popup, we are storing user application object. Shouldn't the check for 'back' be done before these checks and before saving the user_application to db?

**Review by harris-irfan — CHANGES_REQUESTED** [review]

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/views.py:208]: Just a side note, if we need to do backend validations, then it is preferred to override the `clean()` method in the form.

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [inline openedx/adg/lms/applications/views.py:202]: *redirect

File: openedx/adg/lms/applications/models.py
Comment (harris-irfan) [inline openedx/adg/lms/applications/models.py:146]: This is misleading. The function name suggests that both cover letter and cover letter file are being saved, while according to the business logic, either cover_lettter or cover_letter_file is saved.

Plus I would advocate against moving this logic to `models.py` @MunamMubashir-arbisoft. The reason is that moving logic to `models.py` makes more sense if we were to override the `save()` method. Like maybe if we wish to do something pre-save or post-save, every time the application is saved.

In this case, bringing this logic at model level is costly performance wise, because two separate calls are made to the save method in `CoverLetterView.post()`. A separate call is made to the db to first save business line, and then a separate call is made to save cover letter. Calls to DB are costly and should be minimized.

**Review by harris-irfan — CHANGES_REQUESTED** [review]
Test cases are yet to be reviewed.

File: openedx/adg/lms/applications/test/test_views.py
Comment (harris-irfan) [inline openedx/adg/lms/applications/test/test_views.py:487]: We can merge this function with the second last function using parameterization to avoid code repetition, and the logic is also similar. All you need to to do is add another parameter 'cover letter file', and based on that set request.FILES. The rest of the logic stays the same. Parameters are already used quite nicely in the 2nd last function, that too with mocking. Good effort.

**Review by harris-irfan — CHANGES_REQUESTED** [review]
A very good effort on the test cases. I have left a few recommendations to make them even cleaner. Also, do check with @zahid-arbisoft as to why three test cases are failing in circle ci, and if the failing linting job can be fixed or should we approve PRs without that, after just testing for linting locally.

File: openedx/adg/lms/applications/test/test_views.py
Comment (harris-irfan) [inline openedx/adg/lms/applications/test/test_views.py:362]: After merging tests, the code looks lot more clean and readable. For further improvement and consistency, can we also merge these two tests. We can receive `cover_letter_view_get_request` and `cover_letter_view_post_request` as different parameters, in the same test.

**Review by harris-irfan — APPROVED** [review]
Left a single suggestion related to further grouping two related tests. Rest looks fine. Great effort!

Please resolve conflicts before attempting to merge.

Also @zahid-arbisoft, can we merge with failing linting and unit test checks, or should we wait for their fix?

File: openedx/adg/lms/applications/test/test_views.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/test/test_views.py:398]: Maybe we can rename `btn_next` and `template` to something more meaningful.

**Review by MunamMubashir-arbisoft — APPROVED** [review]
Good work 👍🏼

**Pull request merged**
